### ISSUE TITLE
rustc fails to detect duplicate switch cases due to integer overflow

### ISSUE THREAD

**Issue by farcaller** [rust-lang/rust#13727]
A small app to demonstrate the problem:
```rust
fn test(val: u8) {
  match val {
    256 => print!("0b1110\n"),
    512 => print!("0b1111\n"),
    _   => print!("fail\n"),
  }
}

fn main() {
  test(1);
}
```

Compilation fails in llvm:
```
% rustc test.rs
test.rs:3:5: 3:11 warning: literal out of range for its type, #[warn(type_overflow)] on by default
test.rs:3     256 => print!("0b1110\n"),
              ^~~~~~
test.rs:4:5: 4:11 warning: literal out of range for its type, #[warn(type_overflow)] on by default
test.rs:4     512 => print!("0b1111\n"),
              ^~~~~~
Duplicate integer as switch case
  switch i8 %1, label %match_else [
    i8 0, label %match_case
    i8 0, label %match_case3
  ]
i8 0
LLVM ERROR: Broken function found, compilation aborted!
```

Rustc should detect the double label due to overflow.

**Comment by steveklabnik** [comment]
Triage: still reproduces.

**Comment by remram44** [comment]
Still happens. Interestingly, rustc then crashed (Windows, 32b, nightly 157614249 2015-03-01).

**Comment by frewsxcv** [comment]
Visiting for triage: this is still an issue

**Comment by mkpankov** [comment]
Seems this is not due to overflow. I've triggered the bug with this code:

https://is.gd/8DWf5V

**Comment by frewsxcv** [comment]
> Seems this is not due to overflow. I've triggered the bug with this code:

Reduced:

```rust
#[allow(dead_code)]
#[repr(u32)]
enum SomeEnum {
    A = 0x0,
    B = 0x0,
}

#[allow(dead_code)]
fn a(b: SomeEnum) {
    match b {
        SomeEnum::A => (),
        SomeEnum::B => (),
    }
}

fn main() {}
```

```
Duplicate integer as switch case
  switch i32 %1, label %match_else [
    i32 0, label %match_case
    i32 0, label %match_case2
  ], !dbg !22
i32 0
LLVM ERROR: Broken function found, compilation aborted!
playpen: application terminated with error code 1
```

https://is.gd/poTgnO

**Comment by eefriedman** [comment]
I can't reproduce this on nightly; the 256/512 testcase gives an "unreachable pattern" error, and the enum testcase gives "discriminant value `0u32` already exists".

**Comment by frewsxcv** [comment]
I'll create regression tests for this.
